[Sync IV] Integrate Temporal Alignment for ecephys Interfaces and TimeIntervals

CodyCBakerPhD opened the pull request:
Replace #229

## Motivation

Extension of #245 and #244 that integrates the tools developed and tested in those PRs into the common API for all data interfaces. The `BaseDataInterface` defines the methods abstractly but raises `NotImplementedErrors` so that they will have to be defined across child interfaces.

Introduces three synchronization methods for all interfaces:

1) shift of starting time
2) full replacement of timestamps with precisely synchronized timing information
3) inference of synchronized timestamps given the current unsynchronized timestamps within the interface and two references times allowing a mapping from the secondary timing system to the primary one

Still need to:
- [x] Copy the functionality to some behavior interfaces, like DLC used in the tests
- [x] Expand tests
- [x] Open [Sync V], the documentation PR

## How to test the behavior?
Tests are coming together harnessing the synthetic TTL tools established in #245 and #244 - will need to exhaustively check behavior across all supported interfaces, however.

## Checklist

- [X] Have you thoroughly read our [Developer Guide](https://neuroconv.readthedocs.io/en/main/developer_guide.html)?
- [X] Have you ensured the PR description clearly describes the problem and solutions?
- [X] Have you checked to ensure that there aren't other open or previously closed [Pull Requests](https://github.com/catalystneuro/neuroconv/pulls) for the same change?
- [X] If this PR fixes an issue, is the first line of the PR description `fix #XXX` where `XXX` is the issue number?
- [X] Did you update the [CHANGLOG.md](https://github.com/catalystneuro/neuroconv/blob/main/CHANGELOG.md) file on your branch to describe your changes?

bendichter commented on src/neuroconv/basedatainterface.py:
```diff
@@ -46,6 +49,30 @@ def get_metadata(self):
+        timestamps: ArrayType
+            A full vector of timestamps all temporal data in this interface.
+        """
+        if starting_time is not None or timestamps is not None:
+            assert starting_time != timestamps, "Specify either 'starting_time' or 'timestamps', but not both!"
+
+        if starting_time is not None:
+            self.starting_time = starting_time
+        elif timestamps is not None:
+            self.timestamps = np.array(timestamps)
```
I'd actually rather we throw a NotImplemented error here. Otherwise we'll have a seemingly functional `synchronize` method on all children that doesn't actually do anything. It would be find to define a method like `synchronization_validation` that always gets called first and put this code there.

bendichter commented on src/neuroconv/basedatainterface.py:
```diff
@@ -46,6 +49,30 @@ def get_metadata(self):
+            A full vector of timestamps all temporal data in this interface.
+        """
+        if starting_time is not None or timestamps is not None:
+            assert starting_time != timestamps, "Specify either 'starting_time' or 'timestamps', but not both!"
```
why the assert statement? Why not just throw an error? Are you permitting `timestamps == starting_time` as valid? Why?

bendichter commented on src/neuroconv/datainterfaces/ecephys/baserecordingextractorinterface.py:
```diff
@@ -72,6 +71,15 @@ def get_metadata(self):
         return metadata
 
+    def synchronize(self, starting_time: Optional[float] = None, timestamps: Optional[ArrayType] = None):
+        if starting_time is not None or timestamps is not None:
+            assert starting_time != timestamps, "Specify either 'starting_time' or 'timestamps', but not both!"
+
+        if starting_time is not None:
+            self.recording_extractor.set_times(self.recording_extractor.frame_to_times() + starting_time)
+        elif timestamps is not None:
+            self.recording_extractor.set_times(timestamps)
```
There's really no point in defining that abstract method if you're not even going to call it in child classes

bendichter commented on src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py:
```diff
@@ -0,0 +1,137 @@
+            )
+            metadata["Ecephys"].get("Device", list()).append(Device=device)
+
+        # Add groups metadata
+        group_names = self.recording_extractor.get_property("group_name")
+        if self.recording_extractor.get_property("group_name") is None:
+            electrode_groups = [
+                dict(
+                    name="NIDQChannelGroup",
+                    description="A group representing the NIDQ channels.",
+                    location="unknown",
+                    device=device["name"],
+                )
+            ]
+        else:
+            electrode_groups = [
+                dict(
+                    name=group_name,
+                    description="A group representing the NIDQ channels.",
+                    location="unknown",
+                    device=device["name"],
+                )
+                for group_name in set(group_names)
+            ]
+        metadata["Ecephys"].get("ElectrodeGroup", list()).extend(electrode_groups)
+
+        # Electrodes columns descriptions
+        metadata["Ecephys"]["Electrodes"] = [
+            dict(name="group_name", description="Name of the ElectrodeGroup this electrode is a part of."),
+        ]
+
+        metadata["Ecephys"]["ElectricalSeriesSync"] = dict(
+            name="ElectricalSeriesSync", description="Raw acquisition traces for the synchronization (nidq) data."
+        )
+        return metadata
```
https://nwbinspector.readthedocs.io/en/dev/best_practices/time_series.html#subtypes

CodyCBakerPhD replied:
That's something we missed in the NeuroNexus review, then, too.

CodyCBakerPhD replied:
This is quite unfortunate as well; there's a TON of metadata information about the configuration of the NIDQ device and since `TimeSeries` can't accept a `Device` link, I guess I'd be forced to dump all that info into the `TimeSeries.description` instead?

I guess I could also make a separate custom `DynamicTable` to serve as the metadata representation for this stuff - downside is it can't be linked to the series. Other suggestions?

For a more general approach though, it's also extremely intuitive for users (and even devs) to want to use something called `ElectricalSeries` to represent their data that is in units `Volts`; if the issue is less about the `ElectricalSeries` neurodata type, and more about how the `electrodes` table (not yet a neurodata type) is only supposed to specify electrode metadata for neurophysiological-related, then the best solution may simply be to make a proper `ElectrodeTable` type and follow it up with a separate child class that can apply to describing environmental electrodes and all their metadata.

bendichter commented on src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py:
```diff
@@ -0,0 +1,137 @@
+        rising = np.where(diff_dig > 0)[0]
+        falling = np.where(diff_dig < 0)[0]
+
+        ttl_frames = np.concatenate((rising, falling))
```
this line will fail if the recording finishes in the up position

CodyCBakerPhD commented:
Trying to re-trigger tests

CodyCBakerPhD commented:
OK fixed now, that was weird. It wasn't finding the workflow files so the deployment was failing, but no clear reason why it couldn't find them and why it can find them now 🤷‍♂️

bendichter reviewed (commented):
It's really coming together! Looks strong overall. See individual comments for suggestions

bendichter commented on src/neuroconv/datainterfaces/ecephys/baserecordingextractorinterface.py:
```diff
@@ -72,6 +72,19 @@ def get_metadata(self):
+    def synchronize_timestamps(self, timestamps: ArrayType):
+        self.recording_extractor.set_times(times=timestamps)
+
+    def synchronize_with_pulses(self, pulse_timestamps: ArrayType):
+        self.recording_extractor.set_times(
+            times=infer_timestamps_from_pulses(
+                relative_timestamps=self.recording_extractor.get_times(), pulse_timestamps=pulse_timestamps
+            )
+        )
```
```suggestion
    def synchronize_with_pulses(
        self,
        pulse_timestamps_parent: ArrayType,
        pulse_timestamps_child: ArrayType,
        timestamps_child: Optional[ArrayType] = None,
    ):
        if timestamps_child is None:
            timestamps_child = self.recording_extractor.get_times()
        self.recording_extractor.set_times(
            times=infer_timestamps_from_pulses(
                 pulse_timestamps_parent=pulse_timestamps_parent,
                 pulse_timestamps_child=pulse_timestamps_child
                timestamps_child=timestamps_child,
            )
        )
```

CodyCBakerPhD replied:
Yeh, this one is still WIP

CodyCBakerPhD replied:
I wouldn't say parent/child is the best way to refer to the relationship. They're systems acquiring in parallel, but we, or the experimenter, designates one to be the 'primary' timing system which then tracks reference times across the secondary systems.

I've adjusted the names here to reflect this primary/secondary notion, and also how if the timings from the secondary timing streams aren't reflective of exact event occurrences, they become 'reference times'.

CodyCBakerPhD commented on src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py:
```diff
@@ -0,0 +1,144 @@
+            Can also be used to track the duration of an on/off event.
+        """
+        # TODO: consider RAM cost of these operations and implement safer buffering version
+        rising_frames, _ = parse_rising_and_falling_frames_from_ttl(
```
Moving dsicussion to [Sync II]

CodyCBakerPhD commented on src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py:
```diff
@@ -0,0 +1,144 @@
+        ]
+
+        metadata["Ecephys"]["ElectricalSeriesSync"] = dict(
+            name="ElectricalSeriesSync", description="Raw acquisition traces for the synchronization (nidq) data."
```
Done

CodyCBakerPhD commented on src/neuroconv/basedatainterface.py:
```diff
@@ -46,6 +46,58 @@ def get_metadata(self):
+        )
+
+    @abstractmethod
+    def synchronize_with_pulses(self, pulse_timestamps: ArrayType):
```
Renamed the method and input arguments - one thing to note is you may not be parsing any pulses on the NeuroConv side (we've seen several cases where you simply read a vector of the already-calculated reference times) so abstracted away from the notion these are tied to pulses.

Generally we simply have two sets of reference times, those from the primary system we're trying to sync the timestamps of this interface to, and those from the secondary system.

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+        assert_array_equal(x=self.behavior_interface.get_timestamps(), y=expected_behavior_timestamps)
+
+    def test_nidq_interface_synchronized_converter_pipe(self):
+        synchronized_dlc_timestamps = (
+            self.nidq_interface.get_event_starting_times_from_ttl(
+                channel_name="nidq#XA2"  # the channel receiving pulses from the DLC system
+            ),
+        )
+
+        self.dlc_interface.synchronize_timestamps(timestamps=synchronized_dlc_timestamps)
+        self.behavior_interface.synchronize_between_systems(
+            primary_reference_timestamps=self.unsynchronized_secondary_timestamps,
+            secondary_reference_timestamps=synchronized_dlc_timestamps,
+        )
```
This is the primary demonstration of how this kind of workflow would look like in an average conversion script using them

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        # For example, if the synchronization is performed ahead of time and timestamps stored in a CSV or similar file
+        # Then we just need to read in the pulse times instead of having to parse NIDQ signals
+        self.dlc_interface.synchronize_timestamps(timestamps=self.externally_synchronized_timestamps)
+        self.behavior_interface.synchronize_between_systems(
+            primary_reference_timestamps=self.unsynchronized_secondary_timestamps,
+            secondary_reference_timestamps=self.externally_synchronized_timestamps,
+        )
```
This shows how the same setup works in exactly the same way even when the timestamps are given from an external file instead of explicit NIDQ parsing, due to this generalization away from pulses to timestamps (however they might have been parsed originally)

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+        assert_array_equal(x=self.behavior_interface.get_timestamps(), y=expected_behavior_timestamps)
+
+    def test_nidq_interface_synchronized_converter_pipe(self):
+        synchronized_dlc_timestamps = (
+            self.nidq_interface.get_event_starting_times_from_ttl(
+                channel_name="nidq#XA2"  # the channel receiving pulses from the DLC system
+            ),
+        )
+
+        self.dlc_interface.synchronize_timestamps(timestamps=synchronized_dlc_timestamps)
+        self.behavior_interface.synchronize_between_systems(
+            primary_reference_timestamps=synchronized_dlc_timestamps,
+            secondary_reference_timestamps=self.unsynchronized_secondary_timestamps,
```
This is the primary demonstration of how this kind of workflow would look like in an average conversion script using them

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        # For example, if the synchronization is performed ahead of time and timestamps stored in a CSV or similar file
+        # Then we just need to read in the pulse times instead of having to parse NIDQ signals
+        self.dlc_interface.synchronize_timestamps(timestamps=self.externally_synchronized_timestamps)
+        self.behavior_interface.synchronize_between_systems(
+            primary_reference_timestamps=self.externally_synchronized_timestamps,
+            secondary_reference_timestamps=self.unsynchronized_secondary_timestamps,
+        )
```
This shows how the same setup works in exactly the same way even when the timestamps are given from an external file instead of explicit NIDQ parsing, due to this generalization away from pulses to timestamps (however they might have been parsed originally)

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+class TestNIDQSynchronization(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.unsynchronized_dlc_timestamps = np.arange(
+            start=0.0,  # First DLC pulse sent when device turns on
+            stop=17.0,  # Was acquiring for 17 seconds
+            step=1 / 30.0,  # Sent a TTL pulse every on every frame at a rate of 30 fps
+        )
+        cls.synchronized_dlc_timestamps = np.arange(
+            start=3.23,  # DLC device starts acquiring 3.23 seconds after SpikeGLX
+            stop=20.23,
+            step=1 / 30 + 0.00001,  # Synchronization reveals slight drift compared with the SpikeGLX times
+        )
```
For conversational reference to comments above.

Note this is not actually something we've seen with DLC before, I was just using that since it's an established behavioral interface in NeuroConv.

But the setup is taken from work we did with several custom interfaces in the Fee lab: https://github.com/catalystneuro/fee-lab-to-nwb/blob/main/src/fee_lab_to_nwb/happ_ecephys/happmotifinterface.py#L55-L67

Motif timings in a MATLAB file were given in the secondary time system of the audio recorder, and the audio stream had events identified and parsed (by them, timestamps recorded in a separate MATLAB file) so we had to estimate the true motif (and subsequent syllable) times in the primary system given both the IMEC and audio times as reference

CodyCBakerPhD replied:
Swapped from DLC to CsvTimeInterval's since that is actually a use case we've seen, or much closer to it anyway

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+
+        assert_array_equal(x=self.dlc_interface.get_timestamps(), y=synchronized_dlc_timestamps)
+
+        expected_behavior_timestamps = np.array([8.83168, 10.53219, 12.93291])
```
And then these are the expected timestamps for the behavioral events passed that are synchronized between the two systems

CodyCBakerPhD replied:
(The relative unsynchronized times were `[5.6, 7.3, 9.7]`)

CodyCBakerPhD commented on tests/test_on_data/test_synchronization/test_interface_synchronization.py:
```diff
@@ -0,0 +1,149 @@
+
+    def test_nidq_interface_synchronized(self):
+        synchronized_dlc_timestamps = (
+            self.nidq_interface.get_event_starting_times_from_ttl(
```
Could be `get_ttl_times`

CodyCBakerPhD commented on src/neuroconv/datainterfaces/ecephys/baserecordingextractorinterface.py:
```diff
@@ -72,6 +73,26 @@ def get_metadata(self):
+    def synchronize_timestamps(self, synchronized_timestamps: ArrayType):
+        self.recording_extractor.set_times(times=synchronized_timestamps)
+
+    def synchronize_between_systems(
```
```suggestion
    def synchronize_by_interpolation(
```
?

bendichter commented on src/neuroconv/basedatainterface.py:
```diff
@@ -46,6 +47,74 @@ def get_metadata(self):
+    @abstractmethod
+    def synchronize_timestamps(self, synchronized_timestamps: ArrayType):
+        """
+        Replace all timestamps for this interface with those syncrhonized with the primary system.
```
```suggestion
        Replace all timestamps for this interface with those synchronized with the primary system.
```

bendichter commented on src/neuroconv/datainterfaces/ecephys/baserecordingextractorinterface.py:
```diff
@@ -7,7 +7,8 @@
 from ...baseextractorinterface import BaseExtractorInterface
-from ...utils import get_schema_from_hdmf_class, get_base_schema, OptionalFilePathType
+from ...tools.signal_processing import synchronize_timestamps_between_systems
+from ...utils import get_schema_from_hdmf_class, get_base_schema, OptionalFilePathType, ArrayType
```
```suggestion
from ...utils import get_schema_from_hdmf_class, get_base_schema, FilePathType, ArrayType
```

bendichter replied:
I'd like to phase out OptionalFilePathType in favor of Optional[FilePathType]

bendichter commented on src/neuroconv/datainterfaces/ophys/baseimagingextractorinterface.py:
```diff
@@ -1,17 +1,20 @@
     get_schema_from_hdmf_class,
     fill_defaults,
     get_base_schema,
     OptionalFilePathType,
```
```suggestion
    FilePathType,
```

bendichter commented on src/neuroconv/tools/signal_processing.py:
```diff
@@ -40,3 +42,31 @@ def parse_rising_and_falling_frames_from_ttl(trace: np.ndarray, num_bins: int =
     return rising_frames, falling_frames
+
+
+def synchronize_timestamps_between_systems(
+    unsynchronized_timestamps: ArrayType,
+    primary_reference_timestamps: ArrayType,
+    secondary_reference_timestamps: ArrayType,
+) -> np.ndarray:
+    """
+    One method of synchronization used to align timing information across sytems.
+
+    Involves taking timestamps recorded in a secondary system and inferring their corresponding times in a primary
+    system given a pair of reference timestamps between the two systems.
+
+    Parameters
+    ----------
+    unsynchronized_timestamps : ArrayType
+        The timestamps that need to be synchronized into the primary time basis.
+    primary_reference_timestamps : ArrayType
+        A known set of timestamps in the primary system to be used as reference.
+    secondary_reference_timestamps : ArrayType
+        A known set of timestamps in the secondary system to be used as reference.
+
+    Returns
+    -------
+    synchronized_timestamps: numpy.ndarray
+        The interpolation of the unsynchronized_timestamps into the primary reference times.
+    """
+    return np.interp(x=unsynchronized_timestamps, xp=secondary_reference_timestamps, fp=primary_reference_timestamps)
```
Since this is a one-liner, we could just get rid of this function and move `np.interp` to within the methods

CodyCBakerPhD replied:
Agreed

CodyCBakerPhD replied:
(and done)

CodyCBakerPhD commented:
@bendichter We had discussed changing the way we refer to these methods as 'temporal alignment' instead of 'synchronization' to avoid confusing it with the practice of interpolating data like MICrONS did.

Still agree with that? The three methods of this PR could then be

```
align_start_time
align_timestamps
align_by_interpolation
```

bendichter commented:
> @bendichter We had discussed changing the way we refer to these methods as 'temporal alignment' instead of 'synchronization' to avoid confusing it with the practice of interpolating data like MICrONS did.
> 
> Still agree with that? The three methods of this PR could then be
> 
> ```
> align_start_time
> align_timestamps
> align_by_interpolation
> ```

I'm not sure. Let's talk about it in our meeting

CodyCBakerPhD commented on src/neuroconv/basedatainterface.py:
```diff
@@ -46,6 +47,78 @@ def get_metadata(self):
+    @abstractmethod
+    def align_by_interpolation(self, aligned_timestamps: np.ndarray, unaligned_timestamps: np.ndarray):
+        """
+        Interpolate the timestamps of this interface using a mapping from some unaligned time basis to its aligned one.
+
+        Use this method if the unaligned timestamps of the data in this interface are not directly tracked by a primary
+        system, but are known to occur between timestamps that are tracked, then align the timestamps of this interface
+        by interpolating between the two.
+
+        An example could be a metronomic TTL pulse (e.g., every second) from a secondary data stream to the primary
+        timing system; if the time references of this interface are recorded within the relative time of the secondary
+        data stream, then their exact time in the primary system is inferred given the pulse times.
+
+        Must be in units seconds relative to the common 'session_start_time'.
```
@alejoe91 As someone that often deals with this kind of stuff in experiments, how well does this explanation read for you? Any suggestions?

There will be a follow-up to this PR that focuses exclusively on documentation with some real examples, would you be OK with looking at that when the time comes?

CodyCBakerPhD commented on src/neuroconv/datainterfaces/ophys/basesegmentationextractorinterface.py:
```diff
@@ -58,6 +60,26 @@ def get_metadata(self):
+            secondary_reference_timestamps=secondary_reference_timestamps,
+        )
+        self.segmentation_extractor.set_times(times=synchronized_timestamps)
+
```
Turns out it's possible to describe the method purely in terms of the other two alignment methods plus the input arguments, so it's now a part of the base class so no repetition

CodyCBakerPhD commented:
Limiting the scope of this PR to just the `ecephys` interfaces plus the `CsvTimeIntervalsInterface`.

Deeplabcut and other behavioral interfaces will take some extra investigation to see how best to handle timestamps 

Tests are still being polished and debugged, should be ready for review by/on Monday

CodyCBakerPhD commented:
Update: debugging took longer than expected as each new test case revealed something that needed a slight change to some components

As it stands now the features work in all the ways demonstrated by the `test_temporal_alignment` tests, however...

... there is a very mysterious issue that prevents Python 3.7 from properly aligning the behavior test interface using the interpolation method

I've combed through changelogs at the Python and Numpy levels and couldn't find anything that might be the cause

Unsure how much more time to invest in trying to understand the issue given we're expected to drop 3.7 soon anyway

CodyCBakerPhD commented:
- Drop Python 3.7

- Add retrieval for unaligned (or pre-aligned) timestamps vector

- In a follow-up PR, add `starting_time` support to bypass the wasted operation of generating a knowingly regular timestamp vector through SI/ROIExtractors (will require some minor rework to tools functions)

CodyCBakerPhD commented:
Added a method for retrieving the unaltered timestamps, calling it `get_original_timestamps` because we can't actually claim that they are 'unaligned' for any particular session (that's a determination to be made by the user) but we can guarantee that they are the version of the timestamps before calling any sequence of the alignment methods in this PR (which modify an in-place copy of those original timestamps)

CodyCBakerPhD commented on tests/test_on_data/test_temporal_alignment/test_temporal_alignment_methods.py:
```diff
@@ -0,0 +1,424 @@
+
+        self.behavior_interface.align_by_interpolation(
+            aligned_timestamps=inferred_aligned_trial_start_timestamps,
+            unaligned_timestamps=self.unaligned_trial_start_times,
```
```suggestion
            unaligned_timestamps=self.trial_interface.get_original_timestamps(),
```

CodyCBakerPhD commented:
- [ ] add test for align start time

- [ ] investigate expected behavior time values (they seem off)

CodyCBakerPhD commented:
- Adjusted all tests to accurately portray usage of original timestamps when declared as 'unaligned'
- Fixed the behavior event ground truth values that seemed off
  - the problem was not an issue with the functionality, but rather the class setup didn't have the right trial alignment ground truth
- Added the requested functionality for the `TimeIntervalsInterface` to interpolate other columns when one is aligned
  - this is an optional, non-default behavior for safety
  - it is not uncommon to have TTL pulses tracking extra trial columns, so the default behavior is to assume the individual wants to align each column separately
  - it is a convenient method, but there are some things that can go wrong with it, such as how to handle the first or last value when beyond the unaligned range
    - it deals with this by using the `left`/`right` options of `np.interp` using the 'most recent' aligned difference to estimate the gap, but of course this may be wrong in some cases
    - this is likely a case where the linear regression approach could provide a better estimate

CodyCBakerPhD commented:
Also, here's a summary of the changelog, which is almost entirely new line additions...

\+ 91 lines to core `BaseDataInterface` defining these new alignment methods

\+ ~187 lines across each base, behavior, and some special testing interfaces (~17 lines per)  to define temporary methods that each raise `NotImplementedErrors` - this is needed because the new methods are abstract in the base class; and any new data interfaces implemented after the next release will also have to define these methods

\+ 70 lines to `TimeIntervalsInterface` with fully functional and tested implementations of alignment methods
\+ 50 new mock methods needed to test the features of this PR

\+ 527 demonstrative tests of the new functionality (more unit-like tests will be added in #283 and beyond)

bendichter commented on src/neuroconv/basedatainterface.py:
```diff
@@ -46,6 +47,91 @@ def get_metadata(self):
+        """
+        self.align_timestamps(aligned_timestamps=self.get_timestamps() + starting_time)
+
+    def align_by_interpolation(self, unaligned_timestamps: np.ndarray, aligned_timestamps: np.ndarray):
+        """
+        Interpolate the timestamps of this interface using a mapping from some unaligned time basis to its aligned one.
+
+        Use this method if the unaligned timestamps of the data in this interface are not directly tracked by a primary
+        system, but are known to occur between timestamps that are tracked, then align the timestamps of this interface
+        by interpolating between the two.
+
+        An example could be a metronomic TTL pulse (e.g., every second) from a secondary data stream to the primary
+        timing system; if the time references of this interface are recorded within the relative time of the secondary
+        data stream, then their exact time in the primary system is inferred given the pulse times.
+
+        Must be in units seconds relative to the common 'session_start_time'.
+
+        Parameters
+        ----------
+        unaligned_timestamps : numpy.ndarray
+            The timestamps of the unaligned secondary time basis.
+        aligned_timestamps : numpy.ndarray
+            The timestamps aligned to the primary time basis.
+        """
+        self.align_timestamps(
+            aligned_timestamps=np.interp(x=self.get_timestamps(), xp=unaligned_timestamps, fp=aligned_timestamps)
+        )
```
I think the term "timestamps" is a bit overloaded here, because it refers to the sample times of the signal we are trying to align and to the times of the ttl pulses. What do you think about renaming:

unaligned_timestamps -> ttl_sent_times
aligned_timestamps -> ttl_received_times

or something like that

CodyCBakerPhD replied:
`aligned_timestamps` is consistent with the keyword argument for `align_timestamps`, which is consistent with its usage in adjusting the `timestamps` field of the underlying object (when the object is a `TimeSeries`, at least)

Keep in mind the `timestamps` also may not be inferred only through TTL signals - it is also possible to decode the timings via more complicated digital words (compound TTL, each sent at different times and parsed together) or perhaps even some other mechanism (like recording the system time in UNIX seconds and saving it in a CSV)

So I'd say the argument names ought to be agnostic to the 'how' of the timings

If `timestamps` is too long, we can mimic SpikeInterface/ROIExtractors by just referring to it always as `times`

```
get_timestamps -> get_times
align_timestamps -> align_times
unaligned_timestamps -> unaligned_times
aligned_times -> aligned_times
```

in particular, `get_times` would be eventually refactored into that enhanced treatment of `starting_time` + `rate` vs. `timestamps` that will be eventually added to the write tools

CodyCBakerPhD commented:
Merging this to get it into the base - continuing discussion of further improvements in issues #303, #302, and #301
